### PR TITLE
Use WGSL shaders

### PR DESCRIPTION
**Pull request by almarklein**
* [x] Rewrite shaders to WGSL
* [x] Maybe look into first basic version of templating (e.g. jinja2)
* [x] Define uniform structs with numpy dtypes instead of pyshader.
* [x] Auto-generate wgsl struct code from these, warn/fix for alignments?
* [x] Fixed phong lighting.

### CODE REVIEW

File: pygfx/renderers/wgpu/meshrender.py [hunk 1]
```diff
@@ -757,523 +506,339 @@ def helpers(self):
...
-
-    if pos_index < 3:  # or n@u == 0
-        # Just return the same vertex, resulting in degenerate triangles
-        wpos1 = vec4(pos1, 1.0)
-        the_pos = u_stdinfo.projection_transform * u_stdinfo.cam_transform * wpos1
-        the_coord = vec2(0, 0)
-        segment_length = 0.0
+    # Put it together!
+    wgsl = shader.to_string()
+    with open(r"c:/dev/py/tmp.wgsl", "wb") as f:
```
Comment (ivoflipse) [inline pygfx/renderers/wgpu/meshrender.py:541]: Probably not what you want to keep around right?

File: pygfx/renderers/wgpu/_wgpurenderer.py [hunk 2]
```diff
@@ -1039,6 +1041,14 @@ def _update_sampler(self, resource):
         )
         resource._wgpu_sampler = resource.rev, sampler
 
+    def _get_shader_module(self, key, source):
+        """Compile a shader module object, or re-use it from the cache."""
+        # todo: make this work for objects following the ShaderSourceTemplate interface
+        if key not in self._shader_cache:
+            m = self._device.create_shader_module(code=source)
+            self._shader_cache[key] = m
+        return self._shader_cache[key]
```
Comment (Korijn) [inline pygfx/renderers/wgpu/_wgpurenderer.py:1050]: just wondering if you also need to implement code to release the resources?

File: pygfx/renderers/wgpu/meshrender.py (hunk 1 above)
Comment (almarklein) [reply pygfx/renderers/wgpu/meshrender.py:541]: Nice catch!

File: pygfx/renderers/wgpu/_wgpurenderer.py (hunk 2 above)
Comment (almarklein) [reply pygfx/renderers/wgpu/_wgpurenderer.py:1050]: mmm, yeah, in practice the number of shaders used is probably finite, but it does not have to be.

File: pygfx/renderers/wgpu/volumerender.py [hunk 3]
```diff
@@ -43,241 +41,287 @@ def volume_slice_renderer(wobject, render_info):
...
-                func.__annotations__["t_tex"] = tex_anno
-            fragment_shader = fragment_shader.float_version
+            raise ValueError("With VolumeSliceMaterial, geometry needs texcoords")
+        # Sampling type
+        if "norm" in material.map.format or "float" in material.map.format:
+            shader["texture_format"] = "f32"
+            if "unorm" in material.map.format:
+                shader["climcorrection"] = " * 255.0"
+            elif "snorm" in material.map.format:
+                shader["climcorrection"] = " * 255.0 - 128.0"
```
Comment (Korijn) [inline pygfx/renderers/wgpu/volumerender.py:51]: 🙈

**Comment by Korijn** [comment]
Would you mind trying `examples/helpers.py` as a quick check on the coordinate systems?

**Comment by almarklein** [comment]
> Would you mind trying examples/helpers.py as a quick check on the coordinate systems?

I will have a look in a new PR. Will also add some examples to validate coordinates ...

**Pull request merged**